synapse/handlers/federation.py

```diff
@@ -2715,6 +2722,12 @@ def persist_events_and_notify(self, event_and_contexts, backfilled=False):
                 event_and_contexts, backfilled=backfilled
             )
 
+            for (event, context) in event_and_contexts:
+                # If there's an expiry timestamp on the event, schedule its expiry.
+                expiry_ts = event.content.get(EventContentFields.SELF_DESTRUCT_AFTER)
+                if isinstance(expiry_ts, int) and not event.is_state():
+                    yield self._message_handler.maybe_schedule_next_expiry(expiry_ts)
```

Comment:
why not move all of this into the message_handler, so it can be shared between the federation and CS paths?